src/molecules/VoiceRecorder/index.tsx

```diff
@@ -0,0 +1,269 @@
+import { useState } from 'react'
+import styles from './styles.module.css'
+import toast from 'react-hot-toast'
+import config from './config.json'
+
+interface VoiceRecorder {
+  setInputMsg: (msg: string) => void
+  tapToSpeak: boolean
+  includeDiv?: boolean
+}
+
+const VoiceRecorder: React.FC<VoiceRecorder> = ({
+  setInputMsg,
+  tapToSpeak,
+  includeDiv = false,
+}) => {
+  const [mediaRecorder, setMediaRecorder] = useState<MediaRecorder | null>(null)
+  const [userClickedError, setUserClickedError] = useState(false)
+  const [apiCallStatus, setApiCallStatus] = useState('idle')
+
+  const VOICE_MIN_DECIBELS: number = -35
+  const delayBetweenDialogs: number = config.component.delayBetweenDialogs
+  const DIALOG_MAX_LENGTH: number = 60 * 1000
+  let IS_RECORDING: boolean = false
+
+  const startRecording = async () => {
+    IS_RECORDING = true
+    record()
+  }
+
+  const stopRecording = () => {
+    IS_RECORDING = false
+    if (mediaRecorder !== null) {
+      mediaRecorder.stop()
+      setMediaRecorder(null)
+    }
+  }
+
+  function record() {
+    navigator.mediaDevices.getUserMedia({ audio: true }).then((stream) => {
+      //start recording:
+      const recorder = new MediaRecorder(stream)
+      recorder.start()
+      setMediaRecorder(recorder)
+
+      //save audio chunks:
+      const audioChunks: BlobPart[] = []
+      recorder.addEventListener('dataavailable', (event) => {
+        audioChunks.push(event.data)
+      })
+
+      //analysis:
+      const audioContext = new AudioContext()
+      const audioStreamSource = audioContext.createMediaStreamSource(stream)
+      const analyser = audioContext.createAnalyser()
+      analyser.minDecibels = VOICE_MIN_DECIBELS
+      audioStreamSource.connect(analyser)
+      const bufferLength = analyser.frequencyBinCount
+      const domainData = new Uint8Array(bufferLength)
+
+      //loop:
+      let time: Date = new Date()
+      let startTime: number
+      let lastDetectedTime: number = time.getTime()
+      let anySoundDetected: boolean = false
+      const detectSound = () => {
+        //recording stopped by user:
+        if (!IS_RECORDING) return
+
+        time = new Date()
+        const currentTime = time.getTime()
+
+        //time out:
+        if (currentTime > startTime + DIALOG_MAX_LENGTH) {
+          recorder.stop()
+          return
+        }
+
+        //a dialog detected:
+        if (
+          anySoundDetected === true &&
+          currentTime > lastDetectedTime + delayBetweenDialogs
+        ) {
+          recorder.stop()
+          return
+        }
+
+        //check for detection:
+        analyser.getByteFrequencyData(domainData)
+        for (let i = 0; i < bufferLength; i++)
+          if (domainData[i] > 0) {
+            anySoundDetected = true
+            time = new Date()
+            lastDetectedTime = time.getTime()
+          }
+
+        //continue the loop:
+        window.requestAnimationFrame(detectSound)
```

Comment:
use optional chaining with window like window?. because during load time window is not available in nextjs